pennylane/transforms/decompositions/single_qubit_unitary.py

```diff
@@ -58,7 +48,7 @@ def zyz_decomposition(U, wire):
 
     Returns:
         list[qml.Operation]: A ``Rot`` gate on the specified wire that implements ``U``
-        up to a global phase, or an equivalent ``RZ`` gate if ``U`` is diagonal.
```

Comment:
:( I wonder if this leads to any overhead?

Reply:
Ah, I have tried your suggestion from below, and it is indeed possible to work around this. Here's an example, with timings.

Original version (i.e., turn everything into Rot gates, with no conditionals), autograd vs. JAX:

```python
dev = qml.device('default.qubit', wires=1)

@qml.transforms.unitary_to_rot
def circuit(z):
    qml.Hadamard(wires=0)
    qml.QubitUnitary(qml.RZ._matrix(z), wires=0)
    return qml.expval(qml.PauliX(0))

autograd_qnode = qml.QNode(circuit, dev, diff_method="parameter-shift")
grad = qml.grad(autograd_qnode)

z = np.array(0.3, requires_grad=True)

%%timeit
autograd_qnode(z)
1.58 ms ± 8.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%%timeit
grad(z)
6.01 ms ± 22.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Original version with JAX:
```python
z = jnp.array(0.3, dtype=jnp.float64)

jax_qnode = qml.QNode(circuit, dev, interface="jax", diff_method="parameter-shift")
grad = jax.grad(jax_qnode)

%%timeit
jax_qnode(z)
24.2 ms ± 658 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
grad(z)
67.1 ms ± 1.48 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Since there are no conditionals in this version, we can still jit it:

```python
j = jax.jit(jax_qnode)
g = jax.jit(jax.grad(jax_qnode))

%%timeit
j(z)
287 µs ± 1.22 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%%timeit
g(z)
1.98 ms ± 113 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Now add in the conditional statements to turn the diagonal unitaries into `RZ` only. As expected this makes the gradient computation a bit faster.

```python
%%timeit
autograd_qnode(z)
1.45 ms ± 13.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%%timeit
grad(z)
3.37 ms ± 28 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%%timeit
jax_qnode(z)
21.9 ms ± 513 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
grad(z)
48.6 ms ± 589 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

And again, with jit:
```python
j = jax.jit(jax_qnode)
g = jax.jit(jax.grad(jax_qnode))

%%timeit
j(z)
250 µs ± 4.69 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


%%timeit
g(z)
1.06 ms ± 2.77 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

So it seems like it is indeed worth it to do the conditional decomposition, since we can (plus, from a running-on-hardware standpoint, we'll have fewer gates as well).

Reply:
nice!!!!!

Reply:
I must admit I have forgotten the suggestion I made that lead to this improvement 🙃

Reply:
It was just to use `jax.lax.cond` to check whether `RZ` or `Rot` could be used. I have made a mistake, though, in my original implementation in how the conditional was set up. It was choosing `RZ` every time even when something that needed `Rot` was used (this doesn't affect the timings in the above, though, since it is a diagonal operation). I have not been able to fix it yet, there is an issue in how the `wire` parameter is handled. It seems to get passed as a dummy object during the `jit` process, and PennyLane 's` Wires` is *not* happy about that. I think there should be a workaround with partial functions, or something like this...